core/src/ldml/ldml_processor.cpp

```diff
@@ -225,77 +225,84 @@ ldml_processor::process_event(
-      auto cp = state->context();
-      // We're only interested in as much of the context as is a KM_KBP_CT_CHAR.
-      // This will stop at the KM_KBP_CT_MARKER type.
-      uint8_t last_type = KM_KBP_BT_UNKNOWN;
-      for (auto c = cp.rbegin(); c != cp.rend(); c++) {
-        last_type = c->type;
-        if (last_type != KM_KBP_BT_CHAR) {
-          break;
+      {
+        // adapted from kmx_processor.cpp
+
+        /** a copy of the current/changed context, for transform use */
+        ldml::string_list ctxt;
+
+        // Construct a context buffer of all the KM_KBP_BT_CHAR items
+        // Extract the context into 'ctxt' for transforms to process
+        if (!!transforms) {
+          // if no transfors, no reason to do this extraction
+          auto &cp = state->context();
+          // We're only interested in as much of the context as is a KM_KBP_BT_CHAR.
+          uint8_t last_type = KM_KBP_BT_UNKNOWN;
+          for (auto c = cp.rbegin(); c != cp.rend(); c++) {
+            last_type = c->type;
+            if (last_type != KM_KBP_BT_CHAR) {
+              // not a char, get out
+              break;
+            }
+            // extract UTF-32 to 1 or 2 UTF-16 chars in a string
+            km::kbp::kmx::char16_single buf;
+            const int len = km::kbp::kmx::Utf32CharToUtf16(c->character, buf);
+            const std::u16string str(buf.ch, len);
+            ctxt.push_front(str); // prepend to string
+          }
         }
-        km::kbp::kmx::char16_single buf;
-        const int len = km::kbp::kmx::Utf32CharToUtf16(c->character, buf);
-        const std::u16string str(buf.ch, len);
-        ctxt.push_front(str); // prepend to string
-      }
-      if (last_type != KM_KBP_BT_MARKER) {
-        // There was no beginning-of-translation marker.
-        //Add one.
-        state->context().push_marker(0x0);
-        state->actions().push_marker(0x0);
-      }
-      // Look up the key
-      const std::u16string str = keys.lookup(vk, modifier_state);
-      if (str.empty()) {
-        // not found
-        state->actions().commit(); // finish up and
-        return KM_KBP_STATUS_OK; // Nothing to do- no key
-      }
-      const std::u32string str32 = kmx::u16string_to_u32string(str);
-      for(size_t i=0; i<str32.length(); i++) {
-        state->context().push_character(str32[i]);
-        state->actions().push_character(str32[i]);
-      }
-      // add the newly added char
-      ctxt.push_back(str);
-      // Now process transforms
-      std::u16string outputString;
-      // Process the transforms
-      if (!!transforms) {
-
-        // TODO-LDML: unroll ctxt into a str
-        std::u16string ctxtstr;
-        for (size_t i = 0; i < ctxt.size(); i++) {
-          ctxtstr.append(ctxt[i]);
+
+        // Look up the key
+        const std::u16string str = keys.lookup(vk, modifier_state);
+        if (str.empty()) {
+          // not found
+          break; // ----- commit and exit
```

Comment:
there's probably a better structure for this at this point.  but at least, this is updated to have fewer return; points.

Reply:
I don't actually mind `return` -- especially if it reduces depth of algorithm.